src/Evolution/DiscontinuousGalerkin/MortarData.hpp

```diff
@@ -29,117 +29,48 @@ namespace evolution::dg {
 template <size_t Dim>
-class MortarData {
-  using MortarType = std::optional<std::pair<Mesh<Dim - 1>, DataVector>>;
-
- public:
-  /*!
-   * \brief Insert the magnitude of the local face normal, the determinant
-   * of the volume inverse Jacobian, and the determinant of the face Jacobian.
-   * Used for local time stepping with Gauss points.
-   *
-   * The magnitude of the face normal is given by:
-   *
-   * \f{align*}{
-   *  \sqrt{
-   *   \frac{\partial\xi}{\partial x^i} \gamma^{ij}
-   *   \frac{\partial\xi}{\partial x^j}}
-   * \f}
-   *
-   * for a face in the \f$\xi\f$-direction, with inverse spatial metric
-   * \f$\gamma^{ij}\f$.
-   */
-  void insert_local_geometric_quantities(
-      const Scalar<DataVector>& local_volume_det_inv_jacobian,
-      const Scalar<DataVector>& local_face_det_jacobian,
-      const Scalar<DataVector>& local_face_normal_magnitude);
-
-  /*!
-   * \brief Insert the magnitude of the local face normal. Used for local time
-   * stepping with Gauss-Lobatto points.
-   *
-   * The magnitude of the face normal is given by:
-   *
-   * \f{align*}{
-   *  \sqrt{
-   *   \frac{\partial\xi}{\partial x^i} \gamma^{ij}
-   *   \frac{\partial\xi}{\partial x^j}}
-   * \f}
-   *
-   * for a face in the \f$\xi\f$-direction, with inverse spatial metric
-   * \f$\gamma^{ij}\f$.
-   */
-  void insert_local_face_normal_magnitude(
-      const Scalar<DataVector>& local_face_normal_magnitude);
-
-  /*!
-   * \brief Sets the `local_volume_det_inv_jacobian` by setting the DataVector
-   * to point into the `MortarData`'s internal storage.
-   *
-   * \warning The result should never be changed.
-   */
-  void get_local_volume_det_inv_jacobian(
-      gsl::not_null<Scalar<DataVector>*> local_volume_det_inv_jacobian) const;
-
-  /*!
-   * \brief Sets the `local_face_det_jacobian` by setting the DataVector to
-   * point into the `MortarData`'s internal storage.
-   *
-   * \warning The result should never be changed.
-   */
-  void get_local_face_det_jacobian(
-      gsl::not_null<Scalar<DataVector>*> local_face_det_jacobian) const;
-
-  /*!
-   * \brief Sets the `local_face_normal_magnitude` by setting the DataVector to
-   * point into the `MortarData`'s internal storage.
-   *
-   * \warning The result should never be changed.
-   */
-  void get_local_face_normal_magnitude(
-      gsl::not_null<Scalar<DataVector>*> local_face_normal_magnitude) const;
-
-  /// @{
-  /// Data on the mortar
-  auto mortar_data() const
-      -> const std::optional<std::pair<Mesh<Dim - 1>, DataVector>>& {
-    return mortar_data_;
-  }
-
-  auto mortar_data() -> std::optional<std::pair<Mesh<Dim - 1>, DataVector>>& {
-    return mortar_data_;
-  }
-  /// @}
+struct MortarData {
+  std::optional<DataVector> mortar_data{std::nullopt};
+  std::optional<Scalar<DataVector>> face_normal_magnitude{std::nullopt};
+  std::optional<Scalar<DataVector>> face_det_jacobian{std::nullopt};
+  std::optional<Scalar<DataVector>> volume_det_inv_jacobian{std::nullopt};
```

Comment:
These were intentionally 1 allocation because doing allocations on the faces is extremely expensive. They scale with the number of faces and this is really bad. We need some way of keeping it to 1 allocation. Ideally it would be 1 allocation per process per element face, but that'll be harder. I'm not sure what to recommend in terms of design because for nodegroups another optimization is sharing the computation of the Riemann problem between elements on the same node and/or sharing the input in a single memory allocation. From what I've heard this will be even more important on GPUs... I'm not sure if it would be easier to discuss on a call?

Having said that, this impacts Gauss, not GL, though I'm still not convinced that GL is better than G.